### PR TITLE
😬Cleanup loaders

### PR DESCRIPTION
**Pull request by doctrino**
# Description

* Made `build_dir` an init argument of all Loaders, and store it as `resource_build_path` with the appended `folder_name`. This simplifies the initialization of loaders.
* The `find_files` now defaults to the `resource_build_path`. Simplifiying the signauter of `clean_resources` and `deploy_resources`.
* Removed class attribute `_display_name`. Loaders that want to overwrite this just overwrites the `display_name` property.
* Replaced `_is_equal_custom` with a `are_equal` method with a default implementation, that is overwritten for subclasses that needs special compaison. 
* Reordered method in `ResourceLoaders` based on whether the subclass must or can overwrite the methods. To make it easier to understand how to write a resource loader.

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/commands/clean.py [hunk 1]
```diff
@@ -309,13 +306,12 @@ def execute(
...
             if not issubclass(loader_cls, ResourceLoader):
                 continue
-            loader = loader_cls.create_loader(ToolGlobals)
+            loader = loader_cls.create_loader(ToolGlobals, build_dir)
             if type(loader) is DataSetsLoader:
                 self.warn(ToolkitNotSupportedWarning(feature="Dataset clean."))
                 continue
             result = self.clean_resources(
                 loader,
-                build_path / loader_cls.folder_name,
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/commands/clean.py:318]: This is an example for a simplified call to `clean_resources` as the loadrs already knows the build path it is not necessary any more

File: cognite_toolkit/_cdf_tk/commands/deploy.py [hunk 2]
```diff
@@ -192,32 +186,29 @@ def execute(
...
         self,
         loader: ResourceLoader,
-        path: Path,
         ToolGlobals: CDFToolConfig,
         dry_run: bool = False,
         has_done_drop: bool = False,
         has_dropped_data: bool = False,
         verbose: bool = False,
     ) -> ResourceDeployResult | None:
-        loader.build_path = path
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/commands/deploy.py:219]: This looks strange, you are setting an attribute on an input argument (mutating an input argument) with another input argument. 

This is one of the motivation for removing the `path` (which is the build path of the resource`, which is set in the init of the loader.

File: cognite_toolkit/_cdf_tk/commands/clean.py [hunk 3]
```diff
@@ -275,25 +274,23 @@ def _remove_duplicates(
 
 class CleanCommand(CleanBaseCommand):
     def execute(
-        self, ctx: typer.Context, build_dir: str, build_env_name: str, dry_run: bool, include: list[str]
+        self, ctx: typer.Context, build_dir_raw: str, build_env_name: str, dry_run: bool, include: list[str]
```
Comment (ronpal) [inline cognite_toolkit/_cdf_tk/commands/clean.py:277]: Hva menes med "raw"? At det er en str?

Comment (doctrino) [reply cognite_toolkit/_cdf_tk/commands/clean.py:277]: yes, that is correct. With our current configuration of mypy it is not allowed to reuse a variable with another type.

**Pull request merged**